server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java

```diff
@@ -100,27 +138,59 @@ protected void insertMultiTabletsInternally(boolean needCheck) {
       }
     }
     if (insertTabletStatementList.isEmpty()) {
-      return;
+      return null;
     }
 
     InsertMultiTabletsStatement insertMultiTabletsStatement = new InsertMultiTabletsStatement();
     insertMultiTabletsStatement.setInsertTabletStatementList(insertTabletStatementList);
+    return insertMultiTabletsStatement;
+  }
 
+  protected void executeInsertMultiTabletsStatement(
+      InsertMultiTabletsStatement insertMultiTabletsStatement) {
     if (client == null) {
       client = new DataNodeInternalClient(operatorContext.getSessionInfo());
     }
-    TSStatus executionStatus = client.insertTablets(insertMultiTabletsStatement);
-    if (executionStatus.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()
-        && executionStatus.getCode() != TSStatusCode.REDIRECTION_RECOMMEND.getStatusCode()) {
-      String message =
-          String.format(
-              "Error occurred while inserting tablets in SELECT INTO: %s",
-              executionStatus.getMessage());
-      throw new IntoProcessException(message);
+
+    writeOperationFuture =
+        Futures.submit(
+            () -> client.insertTablets(insertMultiTabletsStatement), writeOperationExecutor);
+  }
+
+  /** Return true if the previous write task has done. */
+  protected boolean processWriteOperationFuture() {
+    if (writeOperationFuture == null) {
+      return true;
     }
 
-    for (InsertTabletStatementGenerator generator : insertTabletStatementGenerators) {
-      generator.reset();
+    try {
+      if (!writeOperationFuture.isDone()) {
+        return false;
+      }
```

Comment:
if writeOperationFuture is not done, we should never call next method of into operator, so here if you just want to do a safety check, you should throw a IIlegalStateException.

Reply:
fixed